Source/WebCore/html/HTMLElement.cpp

```diff
@@ -1234,19 +1234,43 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
     return { };
 }
 
-static HTMLElement* topmostPopoverAncestor(Element& element)
-{
-    for (auto& element : lineageOfType<HTMLElement>(element)) {
-        if (element.popoverState() != PopoverState::Auto)
-            continue;
-        if (element.popoverData()->visibilityState() != PopoverVisibilityState::Showing)
+// https://html.spec.whatwg.org/#topmost-popover-ancestor
+// Consider both DOM ancestors and popovers where the given popover was invoked from as ancestors.
+// Use top layer positions to disambiguate the topmost one when both exist.
+static HTMLElement* topmostPopoverAncestor(Element& newPopover)
+{
+    // Store positions to avoid having to do O(n) search for every popover invoker.
+    HashMap<Ref<Element>, size_t> topLayerPositions;
+    size_t i = 0;
+    for (auto& element : newPopover.document().topLayerElements()) {
+        if (!is<HTMLElement>(element) || downcast<HTMLElement>(element.get()).popoverState() != PopoverState::Auto)
             continue;
-        return &element;
+        topLayerPositions.add(element, i);
+        i++;
     }
 
-    // FIXME: Include popover invokers and compare them in top layer order once they are implemented.
+    auto nearestInclusiveOpenPopover = [](Element& candidate) -> HTMLElement* {
```

Comment:
Would be good to add a link to https://html.spec.whatwg.org/#nearest-inclusive-open-popover given that this is what this is implementing.